### PR TITLE
[CI/CD] Setup nodejs/npm environment based on package.json

### PR DESCRIPTION
**Pull request by gnakagaw**
### Summary or Background

This patch add a node/npm setup flow based on `package.json` parameter.

Node 16 has been the default node environment for GitHub Actions runner. But, we have not tested it well. We'll pin the version to Node 14.
https://github.blog/changelog/2021-12-10-github-actions-github-hosted-runners-now-run-node-js-16-by-default/

### Issues or Pull Requests
#2856 

### Checklist

- [-] ~Add a new entry to [CHANGELOG.md](https://github.com/sider/runners/blob/master/CHANGELOG.md) if this change is notable.~

This is a patch to CI configuration. No need to update CHANGELOG.

### Acknowledgement
This patch is based on a @ybiquitous's [comment](https://github.com/sider/runners/pull/2856#discussion_r767703686). Many thanks.

### CODE REVIEW

**Comment by gnakagaw** [comment]
🆗 
https://github.com/sider/runners/runs/4515136502?check_suite_focus=true#step:5:1

**Comment by ybiquitous** [comment]
@gnakagaw This PR seems good to me. 👍🏼 

BTW, it also seems good to update `engines.npm` to `8` in the `package.json` of this repository, because `devon_rex` images use npm 8.1.3. (with another pull request)

https://github.com/sider/runners/blob/7b0b3c1ff2430e9519a529e8f2d04665aabe8a31/package.json#L7

https://github.com/sider/devon_rex/blob/14591028160cd60f21c0faa7c501a6538a44f654/npm/package.json#L3

```
"npm": "8.1.3"
```

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,7 +15,14 @@ jobs:
...
           bundler-cache: true
-      - run: sudo npm install --global npm@7
+      - run: |
+          echo "::set-output name=node::$(jq --raw-output '.engines.node' package.json)"
+          echo "::set-output name=npm::$(jq --raw-output '.engines.npm' package.json)"
+        id: node-versions
+      - uses: actions/setup-node@v2
+        with:
+          node-version: ${{ steps.node-versions.outputs.node }}
+      - run: npm install --global "npm@${{ steps.node-versions.outputs.npm }}"
```
Comment (radomirbosak) [inline .github/workflows/test.yml:25]: The `npm install` command [supports version ranges](https://docs.npmjs.com/cli/v8/commands/npm-install#description) so this is fine.
image

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -15,7 +15,14 @@ jobs:
...
         with:
           bundler-cache: true
-      - run: sudo npm install --global npm@7
+      - run: |
+          echo "::set-output name=node::$(jq --raw-output '.engines.node' package.json)"
+          echo "::set-output name=npm::$(jq --raw-output '.engines.npm' package.json)"
+        id: node-versions
+      - uses: actions/setup-node@v2
+        with:
+          node-version: ${{ steps.node-versions.outputs.node }}
```
Comment (radomirbosak) [inline .github/workflows/test.yml:24]: Is this ok? The `outputs.node` will contain a string like `>=14.16.0`:
```
$ jq --raw-output '.engines.node' package.json
>=14.16.0
```
But the setup-node action syntax description [does not mention](https://github.com/actions/setup-node#supported-version-syntax) support for version ranges.

image

Comment (radomirbosak) [reply .github/workflows/test.yml:24]: Ah, sorry. I tried this on the other repo, but for `runners` this is fine.
```
$ jq --raw-output '.engines.node' package.json
14
```

Comment (radomirbosak) [reply .github/workflows/test.yml:24]: Still, should we prepare for the possibility that the `>=number` syntax appears in this repo, or is it overengineering?

**Review by radomirbosak — APPROVED** [review]
LGTM :+1:

File: .github/workflows/test.yml (hunk 2 above)
Comment (gnakagaw) [reply .github/workflows/test.yml:24]: > Should we prepare for the possibility that the >=number syntax appears in this repo?

IMO, No need it.

**Comment by gnakagaw** [comment]
Thank you.

**Pull request merged**